### DISCUSSION TITLE
Use directory based layout for components

### DISCUSSION THREAD

**Discussion by nerdoc** [tetra-framework/tetra#75] [Ideas]
Components as code are nice for small components with 3 lines of HTML code. But as they grow more complex, they soon get unmaintainable. HTML + JS + CSS code is not well supported in IDEs when trapped in python strings (even if @samwillis wrote an integration for VsCode). PyCharm won't implement this soon, and there are many other glitches like caching, missing IDE helpers (including AI) etc. - syntax highlighting is not enough.

Directory based layouts have some advantages:
* separation of concerns: each language in it's own file: syntax highlighting, IDE helpers are fully available
* proximity of related structures: directories naturally allow grouping of different files.
* easier overview of components, as directories allow easy tree views 

So I propose adding support for component modules as directories with the following structure:

```
    ├── django_app/
    │   ├── components/
    │   │   └── foo_calendar/           # the component
    │   │       ├── __init__.py
    │   │       ├── script.js
    │   │       ├── style.css
    │   │       └── template.html  
    │   ├── apps.py
    │   ├── ...
```

There has to be discussed whether the names of the files should match the component name, like 
```
    ├── django_app/
    │   ├── components/
    │   │   └── foo_calendar/           # the component
    │   │       ├── __init__.py
    │   │       ├── foo_calendar.js
    │   │       ├── foo_calendar.css
    │   │       └── foo_calendar.html  
    │   ├── apps.py
    │   ├── ...
``` 

The first allows faster renaming of components, the latter better finding the components' files using their names.
The `__init__.py` should be the entry point, as it is needed anyway. I tend to implementing the first approach.

I think a management command to create components would be good as well, like

```bash
./manage.py start_component MyLibrary ComponentName
```
#### component-in-a-class vs component-as-dir

Another important question. Should the current component-in-a-class approach be kept?
It would break the API completely, but as Tetra is a relatively new project, things are prone to change, and I am not a fan of keeping old habits - if there is not a really good cause. A good cause would be: >=v1.0.0, stable API, heavily used in production.

Tetra is growing, and I think the components-as-a-directory approach is the much better one.

So I tend to migrate completely to the directory approach.

@gsxdsm - opinions?

**Comment by nerdoc** [comment]
It's not easy to load templates from a folder not listed in settings.TEMPLATES[0]["DIRS"]. This could either be done using a custom Loader, by monkeypatching `settings.TEMPLATES[0]["DIRS"]`, etc.
I think the best approach would be adding a custom loader, that's what [django_components does](https://github.com/EmilStenstrom/django-components/blob/7911daa89ed61132b2a819e74769d748efb3ab03/src/django_components/template_loader.py#L20).

**Comment by gsxdsm** [comment]
Strong supporter of this proposal. Would the server methods go in init.py?

I slightly prefer the second layout with the name of the component for each file type. 

I would propose keeping the existing ability to have the template and script inline with the python as an option, I do think there is a benefit for speed and ease of getting started to have that which can then extend out to separating the components when necessary. 


Would this eliminate the need to register and setup component libraries manually? Essentially the directory becomes a library?

**Reply by nerdoc** [reply]
the second layout is also ok, I just experimented with the first one - As a custom `start_component` mgmt command will be helpful, creating a component should be easy and "automatable".
We surely could keep the existing functionality too.

And yes, this replaces the cumbersome setup of libraries. I have code running here that traverses the components directory, and creates libraries automatically from each folder, as well as putting the Component inheriting class inside of init.py into that library. That already is working.
But I am struggling very badly at the loaders. 

One disadvantage of the loaders is you have to setup the whole thing in settings.py, remove `APPS_DIRS: True`, and maybe tweak staticfiles (like django_components).
As the state lies in the client in Tetra, it seems that when calling the server using a button and the state is reloaded, Tetra uses the Node unpickle function to find the loader again and load the template. This should be working automatically when using a custom loader.

**Reply by gsxdsm** [reply]
I can help take a look at the loaders. Is there a branch with the latest ?

**Reply by gsxdsm** [reply]
And yeah if the first option is working it's totally fine. I think either works and the new command sounds great.

**Reply by nerdoc** [reply]
Just broke my branch again and need some time to fix it again... I'll post it later in the evening today.

**Reply by nerdoc** [reply]
Ok, I pushed to the [components_as_dirs](https://github.com/tetra-framework/tetra/tree/components_as_dirs) branch, even if it is not working - release often, release early.

don't forget to add the custom loader to your settings:
```
            "loaders": [
                "tetra.loaders.components_directories.Loader",
                "django.template.loaders.app_directories.Loader",
                # Fallback to default app directories loader
            ],
```
or use the extended version django_components suggests:
```
            "loaders": [
                (
                    "django.template.loaders.cached.Loader",
                    [
                        "django.template.loaders.filesystem.Loader",
                        "django.template.loaders.app_directories.Loader",
                        "tetra.loaders.components_directories.Loader",
                    ],
                )
            ],
```

**Reply by nerdoc** [reply]
I will continue pushing to this branch, so pull (& push/PR) often please ;-)

**Reply by gsxdsm** [reply]
Okay perfect I'll pull this down tonight and start helping!

**Reply by nerdoc** [reply]
The Library part should be working. Directories in `<app>/components` are treated and registered as libraries automatically.

But I get an error:
```
  File ".../tetra/tetra/components/base.py", line 115, in make_template
    raise ComponentException(
tetra.components.base.ComponentException: Template file '.../tetra/demosite/demo/components/default/reactive_search/reactive_search.html' not found for component ReactiveSearch.
```
although the file exists at the file system.

**Reply by nerdoc** [reply]
I found a way of creating a FileSystemLoader and Engine on the fly to load the template file into memory.

While this still does not work because Tetra needs to know the app of a template (which could be easily added manually), it may be horribly slow, as I doubt it will use Django's template caching system.

But it is an approach that goes without changing the Django settings (no "loaders" section necessary). Please pull.

**Reply by nerdoc** [reply]
At least if you write the full component names like `{% demo.default.ToDo / %}`, it works now. This is a foundation on which we can build.

**Comment by gsxdsm** [comment]
Loving the new approach! Works very well - 

Here is a PR that makes the auto rebuild work for template files found in the component directories - https://github.com/tetra-framework/tetra/pull/77

**Reply by nerdoc** [reply]
Works very well, thank you. I partially had to fix a bug that I had introduced when importing faulty components (and which destroyed the pytests), so I merged your commit locally in my `components_as_dirs` branch and pushed it again. There are still a few things to consider in that branch before merging into main, I keep track of it in https://github.com/tetra-framework/tetra/issues/76.

But thanks for your great patch.

One thing: you are just watching html files. shouldn't css/js files too be watched? Until now, when part of a .py file in a Component class, the rebuild was triggered automatically, but I think this should be done too when these files are changed. my main problem here is: I still am not good in frontend parts and bundling. So the esbuild part still has to be transformed to the component-as-directory approach.

If you could help here too, this would be greatly appreciated... Until now, components are working from directories, but only the .py files are used, the others are ignored...

I'm afraid I don't have time for this in the next 2 weeks. Any help/hint in the `components_as_dirs` branch welcome.

**Reply by gsxdsm** [reply]
Perfect. Yes I was just working on getting the css and js parts ready. I can definitely work on the esbuild changes to make the new approach ready to go and then spend some time testing. Eta will be this weekend.

**Reply by gsxdsm** [reply]
JS/CSS file support has been added.

**Reply by nerdoc** [reply]
Looks great!
